docs/source/format/CanonicalExtensions.rst

```diff
@@ -544,6 +544,49 @@ Primitive Type Mappings
 | UUID extension type  | UUID                   |
 +----------------------+------------------------+
 
+.. _timestamp_with_offset_extension:
+
+Timestamp With Offset
+=============
+This type represents a timestamp column that stores potentially different timezone offsets per value. The timestamp is stored in UTC alongside the original timezone offset in minutes.
+This extension type is intended to be compatible with ANSI SQL's ``TIMESTAMP WITH TIME ZONE``, which is supported by multiple database engines.
+
+* Extension name: ``arrow.timestamp_with_offset``.
+
+* The storage type of the extension is a ``Struct`` with 2 fields, in order:
+
+  * ``timestamp``: a preferably non-nullable ``Timestamp(time_unit, "UTC")``, where ``time_unit`` is any Arrow ``TimeUnit`` (s, ms, us or ns).
+
+  * ``offset_minutes``: a preferably non-nullable signed 16-bit integer (``Int16``) representing the offset in minutes from the UTC timezone. Negative offsets represent time zones west of UTC, while positive offsets represent east. Offsets range from -779 (-12:59) to +780 (+13:00).
+
+* Extension type parameters:
+
+  * ``time_unit``: the time-unit of each of the stored UTC timestamps.
+
+* Description of the serialization:
+
+  Extension metadata is an empty string.
+
+.. note::
+
+   It is also *permissible* for the ``offset_minutes`` field to be dictionary-encoded or run-end-encoded.
+
+.. note::
+
+   It is also *permissible* for ``timestamp`` and ``offset_minutes`` to be nullable, but not recommended.
+
+   If ``timestamp`` is nullable and a value is found to be null, then the whole ``TimestampWithOffset`` value should be interpreted as null. One way of achieving this is to drop ``timestamp``'s validity buffer (V1) and replace the top-level struct's validity buffer (V2) with the result of ``V1 AND V2``.
+
+   If ``offset_minutes`` is nullable and a value is found to be null, then this value should be interpreted as if the offset value were were zero.
```

Comment:
This makes consumption of these data more complicated. Why not mandate that struct fields are always non-null?

Reply:
I think many operations will care only about the timestamp field and having the validity buffer there helps them simply delegate to functions on the `Timestamp` type. At the same time, many function only look at the top-level validity buffer (i.e. the validity of the struct). So during instantiation it would make sense to share the validity buffer if there is one.